internal/blocksync/reactor.go

```diff
@@ -518,91 +517,135 @@ func (r *Reactor) poolRoutine(ctx context.Context, stateSynced bool, blockSyncCh
 			//
 			// TODO: Uncouple from request routine.
 
-			// see if there are any blocks to sync
-			first, second, extCommit := r.pool.PeekTwoBlocks()
-			if first == nil || second == nil || extCommit == nil {
-				if first != nil && extCommit == nil {
+			newBlock, verifyBlock, extCommit := r.pool.PeekTwoBlocks()
+
+			if newBlock == nil || verifyBlock == nil || extCommit == nil {
+				if newBlock != nil && extCommit == nil {
 					// See https://github.com/tendermint/tendermint/pull/8433#discussion_r866790631
-					panic(fmt.Errorf("peeked first block without extended commit at height %d - possible node store corruption", first.Height))
+					panic(fmt.Errorf("peeked first block without extended commit at height %d - possible node store corruption", newBlock.Height))
 				}
 				// we need all to sync the first block
+
 				continue
 			} else {
-				// try again quickly next loop
 				didProcessCh <- struct{}{}
 			}
 
-			firstParts, err := first.MakePartSet(types.BlockPartSizeBytes)
-			if err != nil {
-				r.logger.Error("failed to make ",
-					"height", first.Height,
-					"err", err.Error())
+			newBlockParts, err2 := newBlock.MakePartSet(types.BlockPartSizeBytes)
+			if err2 != nil {
+				r.logger.Error("failed to make block at ",
+					"height", newBlock.Height,
+					"err", err2.Error())
 				return
 			}
 
 			var (
-				firstPartSetHeader = firstParts.Header()
-				firstID            = types.BlockID{Hash: first.Hash(), PartSetHeader: firstPartSetHeader}
+				newBlockPartSetHeader = newBlockParts.Header()
+				newBlockID            = types.BlockID{Hash: newBlock.Hash(), PartSetHeader: newBlockPartSetHeader}
 			)
 
-			// Finally, verify the first block using the second's commit.
-			//
-			// NOTE: We can probably make this more efficient, but note that calling
-			// first.Hash() doesn't verify the tx contents, so MakePartSet() is
-			// currently necessary.
-			// TODO(sergio): Should we also validate against the extended commit?
-			err = state.Validators.VerifyCommitLight(chainID, firstID, first.Height, second.LastCommit)
-
-			if err == nil {
-				// validate the block before we persist it
-				err = r.blockExec.ValidateBlock(ctx, state, first)
+			// TODO(sergio, jmalicevic): Should we also validate against the extended commit?
+			if r.lastTrustedBlock == nil || r.lastTrustedBlock.block == nil {
+				if r.initialState.LastBlockHeight != 0 {
+					r.lastTrustedBlock = &TrustedBlockData{r.store.LoadBlock(r.initialState.LastBlockHeight), r.store.LoadSeenCommit()}
+					if r.lastTrustedBlock == nil {
```

Comment:
But this can never happen, can it?
Did you rather mean `r.lastTrustedBlock.block`?

Reply:
Yes, you are right.